src/api/instances/query/get.ts

```diff
@@ -6,39 +6,110 @@ import { instanceKeys } from "..";
 import { useApiKey } from "~/util/store/apiKey";
 import { useNamespace } from "~/util/store/namespace";
 
+export const triggerValues = ["api", "cloudevent", "instance", "cron"] as const;
+
+export const statusValues = [
+  "pending",
+  "complete",
+  "cancelled",
+  "failed",
+] as const;
+
+export type TriggerValue = (typeof triggerValues)[number];
+export type StatusValue = (typeof statusValues)[number];
+
+export type FiltersObj = {
+  AS?: { type: "CONTAINS"; value: string };
+  STATUS?: {
+    type: "MATCH";
+    value: StatusValue;
+  };
+  TRIGGER?: {
+    type: "MATCH";
+    value: TriggerValue;
+  };
+  AFTER?: { type: "AFTER"; value: Date };
+  BEFORE?: { type: "BEFORE"; value: Date };
+};
+
+export const getFilterQuery = (filters: FiltersObj) => {
+  let query = "";
+  const filterFields = Object.keys(filters) as Array<keyof FiltersObj>;
+
+  filterFields.forEach((field) => {
+    const filterItem = filters[field];
+
+    // Without the guard, TS thinks filterItem may be undefined
+    if (!filterItem) {
+      return console.error("filterItem is not defined");
+    }
+
+    let queryField: string;
+    let queryValue: string;
+
+    if (field === "AFTER" || field === "BEFORE") {
+      const date = filters[field]?.value;
+      if (!date) {
+        throw new Error("date is not defined in date filter");
+      }
+      queryField = "CREATED";
+      queryValue = date.toISOString();
+    } else {
+      const value = filters[field]?.value;
+      if (!value) {
+        throw new Error("filter value is not defined");
+      }
+      queryField = field;
+      queryValue = value;
+    }
+
+    query = query.concat(
+      `&filter.field=${queryField}&filter.type=${filterItem.type}&filter.val=${queryValue}`
+    );
+  });
+
+  return query;
+};
+
 export const getInstances = apiFactory({
   url: ({
     namespace,
     baseUrl,
     limit,
     offset,
+    filter,
   }: {
     baseUrl?: string;
     namespace: string;
     limit: number;
     offset: number;
+    filter?: string;
```

Comment:
Just a thought that came into my mind. What do you think about changing this type from `string` to `FiltersObj` and use `getFilterQuery` inside the `getInstances`. This way `getInstances` would expose a more verbose type and all its consumers don't have to care about how to generate this string.

Reply:
For the query key you should be able to [pass the object](https://tanstack.com/query/v4/docs/react/guides/query-keys#query-keys-are-hashed-deterministically) as it is without needing to call `getFilterQuery` at all. As far as I understand, every change in the filterobject will generate a different filter query string. 

We just have to be aware that
`const filterObj = { "key" : "value", anotherKey: undefined}` and `const filterObj = { "key" : "value" } ` would share the same cache key. But that should not be a problem.